### DISCUSSION TITLE
Tun device seems to randomly change upon connection?

### DISCUSSION THREAD

**Discussion by HybridZach** [OpenVPN/openvpn3-linux#158] [Q&A - Our support arena] [answered]
Been using openvpn3 for a good while now and its been fine, however lately I have noticed something strange. 
When I connect (and this seems to only happen when I initially connect), it will connect fine on tun0, but sometimes work for a bit (like I will be able to open websites and have traffic flow (other times it will not even let me communicate because it will be in the process of switching to tun1 already)) but then seemingly randomly it will change it to tun1. This poses an issue for me (and it could be due to my firewall set up this is happening in the first instance). I can only communicate over tun0 because of the way I have my firewall set up, when it switches randomly to tun1 my firewall blocks the communication. 

Is there a way to help diagnose why this is happening, and or an intended feature of some kind?

I have tried to make sure its not a conflict, like make sure there is no old connections on tun0 still, but its not that, it just seems to randomly decide to switch to tun1 most of the time when connecting. 

I am using the command `openvpn3 sessions-start --config #CONFIGHERE` is there a way to force tun0 only?

If I connect and it switches to tun1, using `openvpn3 session-manage -I tun1 -D` then reconnecting will always make it work again on tun0 it seems every time, so I think something strange is happening here I am not aware of.

**Comment by dsommers** [comment, marked as answer]
Hi,

This is an expected behaviour, actually.  It is the only way to ensure network traffic expected to be routed via the VPN does not leak when there are connectivity issues with the VPN link.  The old approach was to tear down the tun interface, create it again and then setup up IP addresses and routing.  This had the potential of leaking network traffic via your default gateway in the time window the VPN IP address and routes were not present on your system; they disappears when the network interface is removed.

In OpenVPN 3, when there is a connectivity issue and a full reconnect is required, a new tun interface is created and configured and then the old interface is removed.  The problem here with this approach is that the new tun interface cannot have the same interface name as the old one.  Hence the device name change.

I would recommend you to use `--dev tunA` in your configuration file (or if you want to go more fancy, use `--dev vpnA --dev-type tun`).  This should create an interface `tunA0` and `tunA1`.  Your firewall rules, like iptables, can then use wildcards in the interface name ... like `-i tunA+`.

However, it is generally considered even better and more flexible to not be dependent on device names themselves, but rather do the firewall filtering based on subnets instead.  But I do see that there are use cases where this might not be as easy - like when forwarding traffic from the public Internet.

**Reply by HybridZach** [reply]
> In OpenVPN 3, when there is a connectivity issue and a full reconnect is required, a new tun interface is created and configured and then the old interface is removed. The problem here with this approach is that the new tun interface cannot have the same interface name as the old one. Hence the device name change.

Firstly thank you for your time and replies.
Correct me if I am wrong here, so are you saying it remembers using tun0? IE. I connect to tun0 fine, use it a bunch, disconnect shutdown my machine, then when it comes time to connect again the openvpn3 remembers it last used tun0 so will use tun1? 

And so will using `--dev tun0` create security issues? since I use UFW it only allows information via the IP and tun0 of the VPN.
How do I go about placing this into my config file, I am using .ovpn config that the vpn provider provided?

I assume it will be just fine if I just allow tun1 as well in my firewall.
Will it ever go beyond tun1, after tun1 will it use tun0 

Haha sorry if this is a bunch of noob questions, just want to make sure I am fully understanding.
It also matters because I am working on a GUI I spoke to you about still, (it works but once its in a more stable point I will post it on github and there will probably be a lot of improvement to be made). I have not yet implemented the API (as that is a little over my head atm and I am pretty new to coding anyways).

**Reply by dsommers** [reply]
Lets try to clarify a bit.

So, when a full reconnect happens with OpenVPN 3 Linux, it keeps the initial `tun0` around, even though it is not capable to pass any data through that interface (the link is "dead").  Any traffic intended to go via the VPN tunnel will just receive timeouts (typically TCP packets) or getting dropped (typically UDP packets).  OpenVPN 3 Linux now creates a new `tun` interface, so if `tun1` is available, it will use that (it will dynamically try to find the next available interface number).   Then `tun1` is configured, with the VPN IP address and routes, which would typically be the same as on `tun0`.  When `tun0` is then removed, the network traffic now coming will automatically be routed via the new `tun1` interface instead.

If `tun0` would be removed first, like on OpenVPN 2.x, any traffic intended for the VPN would be sent directly to your default gateway - which might be outside of your control, which again could leak network traffic out of your control which was not intended to be sent outside the VPN.

Using `--dev tun0` in your config will not make much difference.  There will be a new interface created.  I would expect that either `tun1` or `tun01` would be created.  That's why I suggested using a non-numerical value (`tunA`), as then the numbering would come on top of this "prefix".  You will need to use wildcard for network devices.

From what I can grasp from the UFW docs, you don't need to provide the interface name: https://help.ubuntu.com/community/UFW :

      sudo ufw allow from <ip address>
      sudo ufw allow from <ip range>/<prefix>
      sudo ufw allow from <target> to <destination> port <port number>
      sudo ufw allow from <target> to <destination> port <port number> proto <protocol name>

**Reply by dsommers** [reply]
btw!  Looking fwd to look at your GUI!  Feel free to reach out if you wonder about how to solve things there.

**Reply by HybridZach** [reply]
Okay so today I just tried to connect and it kept switching between tun0 and tun1.
With varying degrees of working, it would allow me to ping or go to certain sites before hanging and eventually switching between tun0 to tun1 then again it would work for a moment then hang and switch back to tun0. Is this intended because I am sure its not? 

When you say full reconnect do you mean as an example if something happens and it has to reconnect or I disconnect and reconnect in the same login session. Because when I say I have issues connecting initially I mean after a few days I start my machine like today and connect.

The only way to get it stable is to disconnect from tun1 or tun0 with `-I tunx -D` then reconnect again and it then stays with tun0.

Previously I only allowed tun0 on UFW and it with openvpn3 sessions list when ever it said tun1 it would say connected but not work so I think the firewall is working as intended, I allowed tun1 via UFW now it now works when it says tun1 but I have the issue now as described above. I set it up to allow just tun0 and the ip and ports of the VPN so its strict, it only allows via tun1 interface. 

The wildcard for a numerical tun device is a good idea, though are you sure its meant to keep switching between tun0 and 1 I am confused.

And the GUI is ready, simple but works, I just need to change the calls to api calls vs system. Once I solve this issue I will clean up some stuff then upload to github it would be awesome to have you help implement the API and release it for others to use, it would be my first real GUI program in python so its probably messy but it works :P haha.

**Reply by dsommers** [reply]
What does your log file say when this happens (`openvpn3-admin log-service --log-level 6`, increase the log level of your VPN session: `openvpn3 session-manage --config CONFIG_NAME --log-level=6` and watch the logs: `openvpn3 log --config CONFIG_NAME`)

The most common reason why this happens is that the TCP/UDP connection to the server is broken for whatever reason. The client will often try to send some internal ping messages to the server (the `--keepalive` option in the client config or pushed by server as `ping` and `ping-restart`).  If the client does not receive any ping responses, it considers the link dead after `ping-restart` seconds and will then initiate a completely new connection to restore connectivity.

Reasons this is happening is often related to unstable Internet connection or a router somewhere which breaks longer lasting TCP/UDP connections, where "longer" can in the worst times be closer to minutes than hours.

**Reply by HybridZach** [reply]
Thank you again for your time.
Of course it doesn't happen now LOL. 
I just restarted my machine to try reproduce the issue, and its not doing it right now, so I will have to wait till it starts to happen again and try use the logs then. I will update this discussion when I get around to the issue showing itself once more.

> The most common reason why this happens is that the TCP/UDP connection to the server is broken for whatever reason.

I usually always disconnect properly before shutting down my machine, but it could be perhaps something on the end of the VPN provider IDK. I use `openvpn3 session-manage -I tunX -D` and I have started using `--cleanup` also. 

> Reasons this is happening is often related to unstable Internet connection or a router somewhere which breaks longer lasting 
TCP/UDP connections, where "longer" can in the worst times be closer to minutes than hours.

Do you mean unstable on my end? my wifi is usually very good, and I have fibre internet.

**Reply by HybridZach** [reply]
For now the logs only say...
`Attaching to session /net/openvpn/v3/sessions/XXXXXX`

And then once disconnected...
`2022-10-29 23:38:30 Client INFO: Stopping connection
2022-10-29 23:38:30 [STATUS] Connection, Client disconnecting
2022-10-29 23:38:30 [STATUS] Connection, Client disconnected
2022-10-29 23:38:30 Client INFO: Disconnected
2022-10-29 23:38:30 [STATUS] Connection, Client process exited
Session closed`

**Reply by HybridZach** [reply]
Just a quick update, I have tried a bunch of times to reproduce this issue, for some reason it is just not doing it anymore. I just now thought it may be because of me adding tun1 to my firewall, but I removed that and restarted and still, it is connecting just fine. 

However, I will still keep trying to reproduce it and tell you the logs when it does happen next, and I am going to keep tun1 out my firewall for now to see if that triggers it from happening again.

**Reply by HybridZach** [reply]
Update. 
After quite some time of it working perfectly, it finally reproduced the issue I was having before.
I believe this issue only occurs when connecting via my GUI (this uses python `os.popen()`)
I finally was able to grab some logs of what happens when it flakes out.

Initially today it worked right away, and I had data and it was letting me communicate seemingly okay, but after a short while it made everything hang, then suddenly nothing communicates. So then I tired to quickly scramble to enable the log things and caught this stuff for you. (I hope it helps, and I hope its okay me omitting some IPS and stuff).

If I am not mistaken this is the issue? 
`Session invalidated: KEEPALIVE_TIMEOUT`

I have attached the log file - [log.txt](https://github.com/OpenVPN/openvpn3-linux/files/10001714/log.txt)
...I made with all the verbose information.

LMK if there is anything else you need me to do to help diagnose this.

**Reply by dsommers** [reply]
`KEEPALIVE_TIMEOUT` means that the client lost connection to the server; the link between the client and server broke down.  And the connection was re-established by reconnecting to the server.  This will trigger a new tun interface to be created when the reconnection starts.  So this is all just as expected.

**Reply by HybridZach** [reply]
Hmm fair enough. So it is an issue on the VPN providers side I assume?, I have not gotten the issue again today. This time it happened switching VPN servers, so it could be that the server was experiencing issues maybe IDK. Thank you for your continued responses and assistance. If I get it again and its different in the logs I will keep you updated.

**Reply by dsommers** [reply]
Hard to say without seeing server logs.  It can be a server side issue.  And it can be general network issues between your local side and all the way to the server side, so anything from general routing issues, network outages as well as some networks having strict NAT rules which times out too quickly.

**Comment by HybridZach** [comment]
I think its safe to say I no longer experience what ever it was that was causing this issue. Thanks for your time in replying and helping diagnose this. Its safe to say it's an issue on my end not the software.
